ornlneutronimaging/mcpevent2hist
# Refactor core lib into FastSophireadLib

# Description of Pull Request

Refactor the core computing library into `FastSophiread` for better efficiency.

## Purpose of work


Provide a full computation library, `libFastSophiread.a`, for other software to use during processing tpx3 raw data into events.



## Summary of work


1. Renaming: `FastSophireadLib` -> `FastSophiread`
2. Port clustering and peaking fitting algorithms into `FastSophiread` with minor syntax improvement (using newer c++ standard).
3. Corresponding unit tests are ported and modernized.
4. Add benchmark application for
   - raw -> hits (single thread, multi-thread via tab)
   - hits -> events (single thread)
   - raw -> events (single thread, multi-thread via tab)

## Additional detail of work


The default bucket number in abs method is increased from 4 to 8 (performance degradation is acceptable in this case).

## Testing instructions

1. Use the updated environment file to create brand new dev environment (delete the old one if present)
2. Build the branch.
3. Run `ctest -V --output-on-failure` to see all unit test
4. Run all benchmark apps





On Mac M1 Pro:

```bash
./FastSophireadBenchmarks_raw2events.app data/frames_pinhole_3mm_1s_RESOLUTION_000001.tpx3
File size (bytes): 80071504
Read raw data: 0.05349 s

Single thread processing...
Number of hits: 9933804
Number of events: 308915
Single thread processing: 0.321591 s
Single thread processing speed: 3.08896e+07 hits/s

Multi-thread processing...
Number of hits: 9933804
Multi-thread processing: 0.074181 s
Multi-thread processing speed: 1.33913e+08 hits/s
```

and the library build content
```bash
ls -l FastSophiread/
total 5248
drwxr-xr-x  14 chenzhang  staff     448 Sep  7 16:01 CMakeFiles
-rw-r--r--   1 chenzhang  staff    1049 Sep  7 16:01 CTestTestfile.cmake
-rw-r--r--   1 chenzhang  staff   32437 Sep  7 16:01 Makefile
-rwxr-xr-x   1 chenzhang  staff  291772 Sep  7 16:01 SophireadBenchmarks_hits2events
-rwxr-xr-x   1 chenzhang  staff  297675 Sep  7 16:01 SophireadBenchmarks_raw2events
-rwxr-xr-x   1 chenzhang  staff  234041 Sep  7 16:01 SophireadBenchmarks_raw2hits
-rwxr-xr-x   1 chenzhang  staff  315455 Sep  7 16:01 SophireadTests_abs
-rw-r--r--   1 chenzhang  staff     315 Sep  7 16:01 SophireadTests_abs[1]_include.cmake
-rw-r--r--   1 chenzhang  staff     916 Sep  7 16:01 SophireadTests_abs[1]_tests.cmake
-rwxr-xr-x   1 chenzhang  staff   77108 Sep  7 16:01 SophireadTests_centroid
-rw-r--r--   1 chenzhang  staff     335 Sep  7 16:01 SophireadTests_centroid[1]_include.cmake
-rw-r--r--   1 chenzhang  staff    1452 Sep  7 16:01 SophireadTests_centroid[1]_tests.cmake
-rwxr-xr-x   1 chenzhang  staff  235378 Sep  7 16:01 SophireadTests_diskio
-rw-r--r--   1 chenzhang  staff     327 Sep  7 16:01 SophireadTests_diskio[1]_include.cmake
-rw-r--r--   1 chenzhang  staff     498 Sep  7 16:01 SophireadTests_diskio[1]_tests.cmake
-rwxr-xr-x   1 chenzhang  staff  136504 Sep  7 16:01 SophireadTests_fastgaussian
-rw-r--r--   1 chenzhang  staff     351 Sep  7 16:01 SophireadTests_fastgaussian[1]_include.cmake
-rw-r--r--   1 chenzhang  staff     534 Sep  7 16:01 SophireadTests_fastgaussian[1]_tests.cmake
-rwxr-xr-x   1 chenzhang  staff   82927 Sep  7 16:01 SophireadTests_hit
-rw-r--r--   1 chenzhang  staff     315 Sep  7 16:01 SophireadTests_hit[1]_include.cmake
-rw-r--r--   1 chenzhang  staff    2568 Sep  7 16:01 SophireadTests_hit[1]_tests.cmake
-rwxr-xr-x   1 chenzhang  staff  244160 Sep  7 16:01 SophireadTests_tpx3
-rw-r--r--   1 chenzhang  staff     319 Sep  7 16:01 SophireadTests_tpx3[1]_include.cmake
-rw-r--r--   1 chenzhang  staff    2643 Sep  7 16:01 SophireadTests_tpx3[1]_tests.cmake
-rw-r--r--   1 chenzhang  staff    1214 Sep  7 16:01 cmake_install.cmake
-rw-r--r--   1 chenzhang  staff  655912 Sep  7 16:01 libFastSophiread.a
```
